solana/program_v2/programs/gateway_v2/src/instructions/network/update_gatekeeper.rs

```diff
@@ -0,0 +1,65 @@
+use crate::constants::GATEKEEPER_SEED;
+use crate::state::{Gatekeeper, GatekeeperAuthKey, GatekeeperFees};
+use anchor_lang::prelude::*;
+
+// Runs all the update methods on the passed-in gatekeeper
+pub fn update_gatekeeper(
+    ctx: Context<UpdateGatekeeperAccount>,
+    data: UpdateGatekeeperData,
+) -> Result<()> {
+    let gatekeeper = &mut ctx.accounts.gatekeeper;
+    let authority = &mut ctx.accounts.authority;
+
+    gatekeeper.add_auth_keys(&data, authority)?;
+    gatekeeper.add_fees(&data, authority)?;
+    gatekeeper.set_network(&data, authority)?;
+    gatekeeper.set_staking_account(&data, authority)?;
+
+    Ok(())
+}
+
+#[derive(Accounts, Debug)]
+#[instruction(data: UpdateGatekeeperData)]
+pub struct UpdateGatekeeperAccount<'info> {
+    #[account(
+        mut,
+        realloc = Gatekeeper::size(
+                gatekeeper.token_fees.len() + data.token_fees.add.len() - data.token_fees.remove.len(),
+                gatekeeper.auth_keys.len() + data.auth_keys.add.len() - data.auth_keys.remove.len(),
+        ),
+        realloc::payer = authority,
+        realloc::zero = false,
+        seeds = [GATEKEEPER_SEED, authority.key().as_ref(), gatekeeper.gatekeeper_network.key().as_ref()],
+        bump = gatekeeper.gatekeeper_bump,
+    )]
+    pub gatekeeper: Account<'info, Gatekeeper>,
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Clone, Debug, AnchorDeserialize, AnchorSerialize)]
+pub struct UpdateGatekeeperData {
+    /// The [`GatekeeperNetwork`] this gatekeeper is on
+    pub gatekeeper_network: Option<Pubkey>,
+    /// The staking account of this gatekeeper
+    pub staking_account: Option<Pubkey>,
```

Comment:
I'd also put the Staking account Update as a `maybe`. I think it might be necessary, but then I'd still pass it as an Account and update it if it changed.